icu4c/source/common/rbbirb.cpp

```diff
@@ -154,7 +155,18 @@ RBBIDataHeader *RBBIRuleBuilder::flattenData() {
     int32_t reverseTableSize  = align8(fForwardTable->getSafeTableSize());
     int32_t trieSize          = align8(fSetBuilder->getTrieSize());
     int32_t statusTableSize   = align8(fRuleStatusVals->size() * sizeof(int32_t));
-    int32_t rulesSize         = align8((fStrippedRules.length()+1) * sizeof(UChar));
+
+    int32_t rulesLengthInUTF8 = 0;
+    if (fStrippedRules.indexOf(u'\0') >= 0) {
```

Comment:
A nul is allowed in literals in rbbi rules.

I'm pretty sure that unpaired surrogates will also go through the rule parser as literals, but I don't know whether they fail later in the rule building process - like when building the tries.

Reply:
I propose:
- on reading the rule string, use the stored string length, don't rely on NUL-termination
- on writing the rule string don't check for NUL with error, allow NUL

As follow-up (not in this PR):
- early in the rule parser check for unpaired surrogates (= ill-formed UTF-16) and fail with error if present (they have to be escaped)